From f33d46d5300dbd6cd2b3cdc3e5cd4fc29bde3c03 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Date: Thu, 21 Jan 2021 15:11:18 +0200
Subject: [PATCH 44/78] rtc-s32gen1: Simplify probe callback

Replace platform_get_resource + devm_ioremap_nocache with
devm_platform_ioremap_resource. This simplifies the release of all
acquired resources.

Issue: ALB-5823
Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
---
 drivers/rtc/rtc-s32gen1.c | 62 +++++++++++----------------------------
 1 file changed, 17 insertions(+), 45 deletions(-)

diff --git a/drivers/rtc/rtc-s32gen1.c b/drivers/rtc/rtc-s32gen1.c
index 23120728bcc2..61f7352cf6bb 100644
--- a/drivers/rtc/rtc-s32gen1.c
+++ b/drivers/rtc/rtc-s32gen1.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2020 NXP
+ * Copyright 2019-2021 NXP
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -32,7 +32,6 @@
  * struct rtc_s32gen1_priv - RTC driver private data
  * @rtc_base: rtc base address
  * @dt_irq_id: rtc interrupt id
- * @res: rtc resource
  * @rtc_s32gen1_kobj: sysfs kernel object
  * @rtc_s32gen1_attr: sysfs command attributes
  * @pdev: platform device structure
@@ -44,7 +43,6 @@
 struct rtc_s32gen1_priv {
 	u8 __iomem *rtc_base;
 	unsigned int dt_irq_id;
-	struct resource *res;
 	struct kobject *rtc_s32gen1_kobj;
 	struct kobj_attribute rtc_s32gen1_attr;
 	struct platform_device *pdev;
@@ -386,48 +384,27 @@ static int s32gen1_rtc_probe(struct platform_device *pdev)
 	if (!priv)
 		return -ENOMEM;
 
-	priv->res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
-	if (!priv->res) {
-		dev_err(&pdev->dev, "Failed to get resource");
-		err = -ENOMEM;
-		goto err_platform_get_resource;
+	priv->rtc_base = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(priv->rtc_base)) {
+		dev_err(&pdev->dev, "Failed to map registers\n");
+		return PTR_ERR(priv->rtc_base);
 	}
 
-	priv->rtc_base = devm_ioremap_nocache(&pdev->dev,
-		priv->res->start, (priv->res->end - priv->res->start));
-	if (!priv->rtc_base) {
-		dev_err(&pdev->dev, "Failed to map IO address 0x%016llx\n",
-			priv->res->start);
-		err = -ENOMEM;
-		goto err_ioremap_nocache;
-	}
 	dev_dbg(&pdev->dev, "RTC successfully mapped to 0x%p\n",
 		priv->rtc_base);
 
-	priv->rdev = devm_rtc_device_register(&pdev->dev, "s32gen1_rtc",
-					&s32gen1_rtc_ops, THIS_MODULE);
-	if (IS_ERR_OR_NULL(priv->rdev)) {
-		dev_err(&pdev->dev, "devm_rtc_device_register error %ld\n",
-			PTR_ERR(priv->rdev));
-		err = -ENXIO;
-		goto err_devm_rtc_device_register;
-	}
-
 	err = device_init_wakeup(&pdev->dev, ENABLE_WAKEUP);
 	if (err) {
 		dev_err(&pdev->dev, "device_init_wakeup err %d\n", err);
-		err = -ENXIO;
-		goto err_device_init_wakeup;
+		return -ENXIO;
 	}
 
 	if (s32g_priv_dts_init(pdev, priv)) {
-		err = -EINVAL;
-		goto err_dts_init;
+		return -EINVAL;
 	}
 
 	if (s32gen1_rtc_init(priv)) {
-		err = -EINVAL;
-		goto err_rtc_init;
+		return -EINVAL;
 	}
 
 	priv->pdev = pdev;
@@ -439,23 +416,20 @@ static int s32gen1_rtc_probe(struct platform_device *pdev)
 	if (err) {
 		dev_err(&pdev->dev, "Request interrupt %d failed\n",
 			priv->dt_irq_id);
-		err = -ENXIO;
-		goto err_devm_request_irq;
+		return -ENXIO;
 	}
 
 	print_rtc(pdev);
 
-	return 0;
+	priv->rdev = devm_rtc_device_register(&pdev->dev, "s32gen1_rtc",
+					&s32gen1_rtc_ops, THIS_MODULE);
+	if (IS_ERR_OR_NULL(priv->rdev)) {
+		dev_err(&pdev->dev, "devm_rtc_device_register error %ld\n",
+			PTR_ERR(priv->rdev));
+		return -ENXIO;
+	}
 
-err_devm_request_irq:
-err_rtc_init:
-err_dts_init:
-err_device_init_wakeup:
-err_devm_rtc_device_register:
-err_ioremap_nocache:
-	release_resource(priv->res);
-err_platform_get_resource:
-	return err;
+	return 0;
 }
 
 static int s32gen1_rtc_remove(struct platform_device *pdev)
@@ -466,8 +440,6 @@ static int s32gen1_rtc_remove(struct platform_device *pdev)
 	rtcc_val = ioread32(priv->rtc_base + RTCC_OFFSET);
 	iowrite32(rtcc_val & (~RTCC_CNTEN), priv->rtc_base + RTCC_OFFSET);
 
-	release_resource(priv->res);
-
 	dev_info(&pdev->dev, "Removed successfully\n");
 	return 0;
 }
-- 
2.25.1

